src/forest.c

```diff
@@ -481,6 +481,66 @@ int scoutfs_forest_set_bloom_bits(struct super_block *sb,
 	return ret;
 }
 
+/*
+ * The caller is commiting items in the transaction and has found the
+ * greatest item version amongst them.  We store it in the log_trees root
+ * to send to the server.
+ */
+void scoutfs_forest_set_max_vers(struct super_block *sb, u64 max_vers)
+{
+	DECLARE_FOREST_INFO(sb, finf);
+
+	finf->our_log.max_item_vers = cpu_to_le64(max_vers);
+}
+
+/*
+ * The server is calling during setup to find the greatest item version
+ * amongst all the log tree roots.  They have the authoritative current
+ * super.
+ *
+ * Item versions are only used to compare items in log trees, not in the
+ * main fs tree.  All we have to do is find the greatest version amongst
+ * the log_trees so that new locks will have a write_version greater
+ * than all the items in the log_trees.
+ */
+int scoutfs_forest_get_max_vers(struct super_block *sb,
+				struct scoutfs_super_block *super,
+				u64 *vers)
+{
+	struct scoutfs_log_trees *lt;
+	SCOUTFS_BTREE_ITEM_REF(iref);
+	struct scoutfs_key ltk;
+	int ret;
+
+	scoutfs_key_init_log_trees(&ltk, 0, 0);
+	*vers = 0;
+
+	for (;; scoutfs_key_inc(&ltk)) {
+		ret = scoutfs_btree_next(sb, &super->logs_root, &ltk, &iref);
+		if (ret == 0) {
+			if (iref.val_len == sizeof(struct scoutfs_log_trees)) {
+				ltk = *iref.key;
+				lt = iref.val;
+				*vers = max(*vers,
+					    le64_to_cpu(lt->max_item_vers));
+			} else {
+				ret = -EIO;
+			}
+			scoutfs_btree_put_iref(&iref);
+		}
+		if (ret < 0) {
```

Comment:
you don't want an `else` here?

Reply:
Not here, no, 'cause the case in there where the item value wasn't the expected size assigns -EIO.  (It's a gross pattern that the btree api has had for a while: return a reference to an item instead of copying the item into the caller's buffer.  we should fix that.  In this case, then, we'd have a _next interface variant that would return eio if the found item didn't have a value with the expected length.)